### PR TITLE
SDK: Integrate with BalanceObserver

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-sdk/lib/api/sdk.ts [hunk 1]
```diff
@@ -95,6 +100,14 @@ class UniversalLoginSDK {
     this.deploymentObserver = this.deploymentObserver || new DeploymentObserver(this.blockchainService, this.config!.contractWhiteList);
   }
 
+  async getBalanceObserver(ensName: string) {
+    const walletContractAddress = await this.getWalletContractAddress(ensName);
+    ensureNotNull(walletContractAddress, WalletContractNotDeployed);
+    ensureNotNull(this.config, MissingConfiguration);
+
+    this.balanceObserver = this.balanceObserver || new BalanceObserver(this.balanceChecker, walletContractAddress, ACTIVE_TOKENS);
+  }
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/api/sdk.ts]: If balanceObserver is already created we don't need to check walletContractAddress or config.

```ts
    if(this.balanceObserver) {
        return this.balanceObserver 
    } else {
        const walletContractAddress = await this.getWalletContractAddress(ensName);
        ensureNotNull(walletContractAddress, WalletContractNotDeployed);
        ensureNotNull(this.config, MissingConfiguration);
        this.balanceObserver =  new BalanceObserver(this.balanceChecker, walletContractAddress, ACTIVE_TOKENS);
        return this.balanceObserver;
    }
```

File: universal-login-sdk/lib/api/sdk.ts [hunk 2]
```diff
@@ -95,6 +100,14 @@ class UniversalLoginSDK {
     this.deploymentObserver = this.deploymentObserver || new DeploymentObserver(this.blockchainService, this.config!.contractWhiteList);
   }
 
+  async getBalanceObserver(ensName: string) {
+    const walletContractAddress = await this.getWalletContractAddress(ensName);
+    ensureNotNull(walletContractAddress, WalletContractNotDeployed);
+    ensureNotNull(this.config, MissingConfiguration);
+
+    this.balanceObserver = this.balanceObserver || new BalanceObserver(this.balanceChecker, walletContractAddress, ACTIVE_TOKENS);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: return this.balanceObserver;

File: universal-login-sdk/lib/api/sdk.ts [hunk 3]
```diff
@@ -95,6 +100,14 @@ class UniversalLoginSDK {
     this.deploymentObserver = this.deploymentObserver || new DeploymentObserver(this.blockchainService, this.config!.contractWhiteList);
   }
 
+  async getBalanceObserver(ensName: string) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: `fetchBalanceObserver` (and also `fetchDeploymentObserver`, `fetchDeploymentReadyObserver`)

File: universal-login-sdk/lib/api/sdk.ts [hunk 4]
```diff
@@ -95,6 +100,14 @@ class UniversalLoginSDK {
...
+    const walletContractAddress = await this.getWalletContractAddress(ensName);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: let's just do address

File: universal-login-sdk/lib/api/sdk.ts [hunk 5]
```diff
@@ -177,6 +190,11 @@ class UniversalLoginSDK {
     return this.blockchainObserver.subscribe(eventType, filter, callback);
   }
 
+  subscribeToBalances(callback: Function) {
+    ensureNotNull(this.balanceObserver, BalanceObserverNotCreated);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: ensure -> fetchBalanceObserver()

File: universal-login-sdk/lib/core/utils/ActiveTokens.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import {TokenDetails, ETHER_NATIVE_TOKEN} from '@universal-login/commons';
+
+export const ACTIVE_TOKENS: TokenDetails[] = [
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/utils/ActiveTokens.ts]: Move to configuration

File: universal-login-sdk/lib/api/sdk.ts (hunk 2 above)
Comment (marekkirejczyk) [reply universal-login-sdk/lib/api/sdk.ts]: Let's introduce 
```ts
interface SDKConfig {
  defaultPaymentOptions?: Message,
  observedTokens: []
}
```

By default take tokens from relayer SupportedTokens.

Comment (marekkirejczyk) [reply universal-login-sdk/lib/api/sdk.ts]: Let's use TokenService to fetch tokens data.
TokenService should be moved from wallet to commons? (or sdk if hard)

File: universal-login-sdk/lib/api/sdk.ts (hunk 5 above)
Comment (marekkirejczyk) [reply universal-login-sdk/lib/api/sdk.ts]: fetchTokenDetails()

File: universal-login-sdk/lib/core/utils/errors.ts [hunk 7]
```diff
@@ -97,6 +97,27 @@ export class MissingMessageHash extends NotFound {
...
+}
+export class NotCreated extends SDKError {
+  constructor (message: string, errorType: ErrorType) {
+    super(message, errorType);
+    this.errorType = errorType;
+    Object.setPrototypeOf(this, NotCreated.prototype);
+  }
+}
+
+export class BalanceObserverNotCreated extends NotCreated {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/utils/errors.ts]: Unnecessary

File: universal-login-sdk/lib/core/utils/errors.ts [hunk 8]
```diff
@@ -97,6 +97,27 @@ export class MissingMessageHash extends NotFound {
   }
 }
 
+export class WalletContractNotDeployed extends NotFound {
+  constructor () {
+    super('Wallet Contract has not been deployed', 'WalletContractNotDeployed');
+    Object.setPrototypeOf(this, WalletContractNotDeployed.prototype);
+  }
+}
+export class NotCreated extends SDKError {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/utils/errors.ts]: Unnecessary

File: universal-login-sdk/lib/core/utils/errors.ts [hunk 9]
```diff
@@ -97,6 +97,27 @@ export class MissingMessageHash extends NotFound {
   }
 }
 
+export class WalletContractNotDeployed extends NotFound {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/utils/errors.ts]: InvalidContract

File: universal-login-sdk/lib/api/sdk.ts (hunk 5 above)
Comment (gasiortomasz) [reply universal-login-sdk/lib/api/sdk.ts]: > ensure -> fetchBalanceObserver()
But fetchBalanceObserver receives ensName as an argument. Passing ensName to subscribe is kinda scratchy, no?

File: universal-login-commons/test/integration/ethereum/TokenDetailsService.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import {providers, Wallet} from 'ethers';
+import {createMockProvider, getWallets, deployContract} from 'ethereum-waffle';
+import {expect} from 'chai';
+import {TokenDetailsService} from '../../../lib/integration/ethereum/TokenDetailsService';
+import {ETHER_NATIVE_TOKEN} from '../../../lib/core/constants/constants';
+import MockToken from '../../fixtures/MockToken.json';
+
+describe('INT: TokenDetailsService', () => {
```
Comment (marekkirejczyk) [inline universal-login-commons/test/integration/ethereum/TokenDetailsService.ts]: 👍 Nice tests

File: universal-login-commons/test/integration/ethereum/TokenDetailsService.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    const symbol = await tokenService.getSymbol(mockToken.address);
+    const name = await tokenService.getName(mockToken.address);
+
+    expect(symbol).to.eq('Mock');
+    expect(name).to.eq('MockToken');
+  });
+
+  it('token not deployed', async () => {
+    const notDeployedtokenAddress = '0x000000000000000000000000000000000000DEAD';
+    expect(tokenService.getSymbol(notDeployedtokenAddress)).to.be.eventually.rejectedWith('contract not deployed');
```
Comment (marekkirejczyk) [inline universal-login-commons/test/integration/ethereum/TokenDetailsService.ts]: We should probably repackage into a more informative exception, i.e. `'Invalid contract address: 0x000000000000000000000000000000000000DEAD'`

File: universal-login-commons/lib/integration/ethereum/TokenDetailsService.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import {providers, Contract} from 'ethers';
+import {ETHER_NATIVE_TOKEN} from '../../core/constants/constants';
+
+const tokenAbi = [
+  'function name() public view returns (string)',
+  'function symbol() public view returns (string)'
+];
+
+export class TokenDetailsService {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/integration/ethereum/TokenDetailsService.ts]: Nice abstraction 👍

File: universal-login-sdk/lib/core/utils/errors.ts [hunk 13]
```diff
@@ -47,6 +47,13 @@ export class InvalidAddress extends ValidationFailed {
   }
 }
 
+export class InvalidContract extends ValidationFailed {
+  constructor () {
+    super('Contract is not valid', 'InvalidContract');
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/utils/errors.ts]: To make exception more informative, we should include the contract address.

File: universal-login-sdk/lib/config/SdkConfigDefault.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import {ETHER_NATIVE_TOKEN} from '@universal-login/commons';
+import {SdkConfig} from './SdkConfig';
+import MESSAGE_DEFAULTS from '../core/utils/MessageDefaults';
+
+export const SdkConfigDefault: SdkConfig = {
+  paymentOptions: MESSAGE_DEFAULTS,
+  observedTokens: [
+    {address: ETHER_NATIVE_TOKEN.address}
+  ]
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/config/SdkConfigDefault.ts]: The default should be taken from relayer. Next PR?

**Comment by marekkirejczyk** [comment]
Merging. Smaller improvements to be done in following PRs.

**Comment by marekkirejczyk** [comment]
One more thing to do in next PR, deduplicate TokenService from Wallet:
https://github.com/UniversalLogin/UniversalLoginSDK/blob/1dcdb0a740cae279df7b8792491400965bf04a5f/universal-login-wallet/test/integration/services/TokenService.ts

**Pull request merged**